storage/src/main/java/zipkin2/storage/kafka/KafkaSpanStore.java

```diff
@@ -178,53 +181,74 @@
   static class GetTracesCall extends KafkaStreamsStoreCall<List<List<Span>>> {
     final ReadOnlyKeyValueStore<String, List<Span>> tracesStore;
     final ReadOnlyKeyValueStore<Long, Set<String>> traceIdsByTsStore;
-    final QueryRequest queryRequest;
+    final QueryRequest request;
 
     GetTracesCall(
         ReadOnlyKeyValueStore<String, List<Span>> tracesStore,
         ReadOnlyKeyValueStore<Long, Set<String>> traceIdsByTsStore,
-        QueryRequest queryRequest) {
+        QueryRequest request) {
       this.tracesStore = tracesStore;
       this.traceIdsByTsStore = traceIdsByTsStore;
-      this.queryRequest = queryRequest;
+      this.request = request;
     }
 
     @Override public List<List<Span>> query() {
       List<List<Span>> traces = new ArrayList<>();
       List<String> traceIds = new ArrayList<>();
       // milliseconds to microseconds
-      long from = (queryRequest.endTs() - queryRequest.lookback()) * 1000;
-      long to = queryRequest.endTs() * 1000;
-      // first index
-      KeyValueIterator<Long, Set<String>> spanIds = traceIdsByTsStore.range(from, to);
-      spanIds.forEachRemaining(keyValue -> {
-        for (String traceId : keyValue.value) {
-          if (!traceIds.contains(traceId) && traces.size() < queryRequest.limit()) {
-            List<Span> spans = tracesStore.get(traceId);
-            if (spans != null && queryRequest.test(spans)) { // apply filters
-              traceIds.add(traceId); // adding to check if we have already add it later
-              traces.add(spans);
+      long from = (request.endTs() - request.lookback()) * 1000;
+      long to = request.endTs() * 1000;
+      long checkpoint = to - (30 * 1000 * 1000); // 30 sec before upper bound
+      if (checkpoint <= from) { // do one run
+        try (KeyValueIterator<Long, Set<String>> spanIds = traceIdsByTsStore.range(from, to)) {
+          spanIds.forEachRemaining(keyValue -> {
+            for (String traceId : keyValue.value) {
+              if (!traceIds.contains(traceId)) {
+                List<Span> spans = tracesStore.get(traceId);
+                if (spans != null && request.test(spans)) { // apply filters
+                  traceIds.add(traceId); // adding to check if we have already add it later
+                  traces.add(spans);
+                }
+              }
             }
+          });
+        }
+      } else {
+        while (checkpoint > from && traces.size() < request.limit()) {
```

Comment:
cc @llinder @michaelsembwever this seems familiar though in cassandra I think we try to do async lazy chained calls.. not sure if the `traceIdsByTsStore.range` here is a remote call or not.

Reply:
All stores are local tho

Reply:
Unless we'd support https://github.com/openzipkin/zipkin/issues/2784 where calls could span multiple instances

Reply:
Issue moved to #36